programs/perpetuals/src/instructions/get_lp_token_price.rs

```diff
@@ -0,0 +1,74 @@
+//! GetLpTokenPrice instruction handler
+
+use {
+    crate::{
+        math::{checked_as_f64, checked_float_div, to_token_amount},
+        state::{
+            perpetuals::Perpetuals,
+            pool::{AumCalcMode, Pool},
+        },
+    },
+    anchor_lang::prelude::*,
+    anchor_spl::token::Mint,
+    num_traits::Zero,
+};
+
+#[derive(Accounts)]
+pub struct GetLpTokenPrice<'info> {
+    #[account(
+        seeds = [b"perpetuals"],
+        bump = perpetuals.perpetuals_bump
+    )]
+    pub perpetuals: Box<Account<'info, Perpetuals>>,
+
+    #[account(
+        seeds = [b"pool",
+                 pool.name.as_bytes()],
+        bump = pool.bump
+    )]
+    pub pool: Box<Account<'info, Pool>>,
+
+    #[account(
+        seeds = [b"lp_token_mint",
+                 pool.key().as_ref()],
+        bump = pool.lp_token_bump
+    )]
+    pub lp_token_mint: Box<Account<'info, Mint>>,
+    // remaining accounts:
+    //   pool.tokens.len() custody accounts (read-only, unsigned)
+    //   pool.tokens.len() custody oracles (read-only, unsigned)
+}
+
+#[derive(AnchorSerialize, AnchorDeserialize)]
+pub struct GetLpTokenPriceParams {}
+
+pub fn get_lp_token_price(
+    ctx: Context<GetLpTokenPrice>,
+    _params: &GetLpTokenPriceParams,
+) -> Result<u64> {
+    let aum_usd = ctx.accounts.pool.get_assets_under_management_usd(
+        AumCalcMode::EMA,
+        ctx.remaining_accounts,
+        ctx.accounts.perpetuals.get_time()?,
+    )?;
+
+    msg!("aum_usd: {}", aum_usd);
+
+    let lp_supply = ctx.accounts.lp_token_mint.supply;
+
+    msg!("lp_supply: {}", lp_supply);
+
+    if lp_supply.is_zero() {
+        return Ok(0);
+    }
+
+    let ui_price_usd = checked_float_div(checked_as_f64(aum_usd)?, checked_as_f64(lp_supply)?)?;
```

Comment:
VM doesn't support floating point operations, they are emulated and hence slow. Also less precise for large integers comparing to u128. And you need to account for decimals in lp_supply. I think this should work:
let price_usd = math::checked_decimal_div(
                aum_usd as u128, -(Perpetuals::USD_DECIMALS as i32),
                lp_supply as u128, -(Perpetuals::LP_DECIMALS as i32),
                -(Perpetuals::USD_DECIMALS as i32),
            )?)?;
this would be the price you return, no need to do to_token_amount()

Reply:
> VM doesn't support floating point operations, they are emulated and hence slow.

Interesting I didn't know that, thx 🙏 

Updated the instruction to use `math::checked_decimal_div`.

So I could understand `math::checked_decimal_div` edge cases better I've added some unit tests in `math` file:

```
 *  Executing task: cargo test --package perpetuals --lib -- math::tests --nocapture 

warning: profiles for the non root package will be ignored, specify profiles at the workspace root:
package:   /Users/orex/work/perpetuals/programs/perpetuals/Cargo.toml
workspace: /Users/orex/work/perpetuals/Cargo.toml
    Finished test [unoptimized + debuginfo] target(s) in 0.25s
     Running unittests src/lib.rs (target/debug/deps/perpetuals-930e8997802f4402)

running 1 test
test math::tests::test_checked_decimal_div_ok ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 13 filtered out; finished in 0.00s

 *  Terminal will be reused by tasks, press any key to close it.
 ```

Reply:
Ha I know that I'm stupid I missed it. 

We used to use a fixed point math lib in the past and that made us stupid. Thanks